pkg/kwokctl/runtime/binary/cluster_snapshot.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package binary
+
+import (
+	"context"
+	"os"
+	"strconv"
+
+	"sigs.k8s.io/kwok/pkg/kwokctl/runtime"
+	"sigs.k8s.io/kwok/pkg/kwokctl/utils"
+	"sigs.k8s.io/kwok/pkg/kwokctl/vars"
+)
+
+// SnapshotSave save the snapshot of cluster
+func (c *Cluster) SnapshotSave(ctx context.Context, path string) error {
```

Comment:
There is a lot of commonality for snapshotting + restore sequence, like running `etcdctl snapshot save` and stop etcd, restore, and start

I'm wondering if we can further refactor this commonality out and just leave the difference like running `docker cp` and `docker exec rm`  to the specific runtime that would be even better. Right now the code still looks a little repetitive.

Reply:
This is difficult, there are many details that are different.